packages/experience-builder-sdk/src/blocks/editor/VisualEditorContext.tsx

```diff
@@ -116,28 +116,42 @@ export function VisualEditorContextProvider({
     );
   }, [entityStore, locale]);
 
-  // When the tree was updated, we store the dataSource and
-  // afterward, this effect fetches the respective entities.
-  useEffect(() => {
-    const resolveEntities = async () => {
-      const dataSourceEntityLinks = Object.values(dataSource || {});
-      const entityLinks = [...dataSourceEntityLinks, ...(designComponentsRegistry.values() || [])];
-      const fetchingResponse = entityStore.fetchEntities(entityLinks);
-      // Only update the state and rerender when we're actually fetching something
-      if (fetchingResponse === false) return;
-      setEntitiesFetched(false);
-      // Await until the fetching is done to update the state variable at the right moment
-      await fetchingResponse;
-      setEntitiesFetched(true);
-      console.debug('[exp-builder.sdk] Finish fetching entities', {
+  // Either gets called when dataSource changed or designComponetsRegistry changed (manually)
+  const fetchMissingEntities = useCallback(async () => {
+    if (isFetchingEntities) return;
+    const requiredEntityLinks = [
+      ...Object.values(dataSource),
+      ...designComponentsRegistry.values(),
+    ];
+    const { missingAssetIds, missingEntryIds } =
+      entityStore.getMissingEntityIds(requiredEntityLinks);
+    const areEntitiesFetched = !missingAssetIds.length && !missingEntryIds.length;
```

Comment:
Will this cause any issues if an entity is legitimately missing, e.g. if it has been deleted from the space?

Reply:
Mmmh, you're right that this probably ends up in an infinite loop of fetching.
I will revert my fix here and reduce the PR to the quality of life improvements. His approach might be more stable in that sense.